File: leetcode3/최원준/341. Flatten Nested List Iterator.py
#

'''
1. 아이디어 :
재귀로 풀 수 있다.
nestedList를 순회하면서 숫자면 배열에 추가, 아니라면 nestedList를 재귀로 호출하여 반복.

2. 시간복잡도 :
    O(n), O(1), O(1)

3. 자료구조/알고리즘 :
배열, dfs

'''

# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
#class NestedInteger:
#    def isInteger(self) -> bool:
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        """
#
#    def getInteger(self) -> int:
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        """
#
#    def getList(self) -> [NestedInteger]:
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        """

class NestedIterator:
    def __init__(self, nestedList: [NestedInteger]):
        self.arr = []
        self.idx = 0
        
        def dfs(nestedList):
            for e in nestedList:
                if e.isInteger():
                    self.arr.append(e.getInteger())
                else:
                    dfs(e.getList())
        
        dfs(nestedList)
    
    def next(self) -> int:
        self.idx += 1
        return self.arr[self.idx-1]
    
    def hasNext(self) -> bool:
        return self.idx < len(self.arr)
        

# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())

File: leetcode3/이진희/341_Flatten Nested List Iterator.java
/*

1. 아이디어 : 주어진 인터페이스 활용, 아래 메소드를 완성
              1. public NestedIterator(List<NestedInteger> nestedList) -> nestedList 초기화
              2. public Integer next() -> 다음 수 출력
              3. public boolean hasNext() -> 현재 위치에서 nested list에 다음 Integer가 존재시 return true

              이때, 리스트 원소를 뒤부터 넣어줘야하는데, 원래 순서로 풀으려면, 나중에 들어간 요소가 먼저나와야 하기 때문

2. 시간복잡도 : 1. NestedIterator -> O(N)
                2. Integer next() -> O(1)
                3. boolean hasNext() -> O(N)

3. 자료구조/알고리즘 : Deque(Stack 처럼 활용)

 */

public class NestedIterator implements Iterator<Integer> {
    private Deque<NestedInteger> deque;

    public NestedIterator(List<NestedInteger> nestedList) {
        deque = new ArrayDeque<>();

        for(int i=nestedList.size()-1; i>=0; i--) {
            deque.push(nestedList.get(i));
        }
    }

    @Override
    public Integer next() {
        int num = deque.pop().getInteger();
        return num;
    }

    @Override
    public boolean hasNext() {
        while(!deque.isEmpty()) {
            boolean check = deque.peek().isInteger();

            if(check) return true;
            List<NestedInteger> list = deque.pop().getList();

            for(int i=list.size()-1; i>=0; i--) {
                deque.push(list.get(i));
            }
        }
        return false;
    }
}